File: cpp/leetcode/editor/cn/[230]二叉搜索树中第K小的元素.cpp
#include <vector>
#include "util.h"
#include <map>
#include <stack>
#include <string>
#include <utility>
#include <memory>
#include <iterator>
#include <unordered_map>
#include <algorithm>
#include <numeric>

using namespace std;
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Answer {
public:
    Answer(TreeNode* root,int k):root(root),k(k){}

    int solution() {
        traverse(root);
        return res;
    }

    void traverse(TreeNode *node) {
        if (node == nullptr) return;
        traverse(node->left);

        ++rank;
        if (rank == k){
            res = node->val;
            return;
        } else
            traverse(node->right);

    }
private:
    int res = 0;
    int rank = 0;
    TreeNode* root;
    int k;
};
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        return Answer(root,k).solution();
    }
};
//leetcode submit region end(Prohibit modification and deletion)

File: scala/leetcode_scala/src/main/scala/leetcode/editor/cn/[230]二叉搜索树中第K小的元素.scala
package leetcode.editor.cn

object KthSmallestElementInABst {
//leetcode submit region begin(Prohibit modification and deletion)
  /** Definition for a binary tree node. class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) { var value: Int = _value var left: TreeNode = _left var right: TreeNode =
    * _right }
    */
  object Solution {
    class Answer(val root: TreeNode, val k: Int) {
      var res = 0
      var rank = 0

      def solution:Int =
        traverse(root)
        res
      def traverse(node: TreeNode): Unit =
        if node != null then
          traverse(node.left)
          rank += 1
          if k == rank then res = node.value
          else traverse(node.right)
    }

    def kthSmallest(root: TreeNode, k: Int): Int =
      Answer(root,k).solution

  }
//leetcode submit region end(Prohibit modification and deletion)

}
